pkg/pillar/types/errortime.go

```diff
@@ -10,76 +10,90 @@ import (
-	}
-	etPtr.Error = errStr
-	etPtr.ErrorTime = errorTime
-}
-
-// ClearError removes it
-func (etPtr *ErrorAndTime) ClearError() {
-	etPtr.Error = ""
-	etPtr.ErrorTime = time.Time{}
-}
-
-// HasError returns true if there is an error
-func (etPtr *ErrorAndTime) HasError() bool {
-	return etPtr.Error != ""
-}
+// ErrorSeverity tells the severity type, must be in sync with info api
+type ErrorSeverity int32
+
+const (
+	// ErrorSeverityUnspecified severity unspecified
+	ErrorSeverityUnspecified ErrorSeverity = 0
+	// ErrorSeverityNotice severity notice
+	ErrorSeverityNotice ErrorSeverity = 1
+	// ErrorSeverityWarning severity warning
+	ErrorSeverityWarning ErrorSeverity = 2
+	// ErrorSeverityError severity error
+	ErrorSeverityError ErrorSeverity = 3
+)
 
-// NewErrorAndTime returns instance of ErrorAndTime with the specified values
-func NewErrorAndTime(errStr string, errTime time.Time) ErrorAndTime {
-	return ErrorAndTime{Error: errStr, ErrorTime: errTime}
-}
+// ErrorEntityType contains the entity type, must be in sync with info api
+type ErrorEntityType int32
+
+const (
+	// ErrorEntityUnspecified Entity
+	ErrorEntityUnspecified ErrorEntityType = 0
+	// ErrorEntityBaseOs entity
+	ErrorEntityBaseOs ErrorEntityType = 1
+	// ErrorEntitySystemAdapter Entity
+	ErrorEntitySystemAdapter ErrorEntityType = 2
+	// ErrorEntityVault Entity
+	ErrorEntityVault ErrorEntityType = 3
+	// ErrorEntityAttestation Entity
+	ErrorEntityAttestation ErrorEntityType = 4
+	// ErrorEntityAppInstance Entity
+	ErrorEntityAppInstance ErrorEntityType = 5
+	// ErrorEntityPort Entity
+	ErrorEntityPort ErrorEntityType = 6
+	// ErrorEntityNetwork Entity
+	ErrorEntityNetwork ErrorEntityType = 7
+	// ErrorEntityNetworkInstance Entity
+	ErrorEntityNetworkInstance ErrorEntityType = 8
+	// ErrorEntityContentTree Entity
+	ErrorEntityContentTree ErrorEntityType = 9
+	// ErrorEntityContentBlob Entity
+	ErrorEntityContentBlob ErrorEntityType = 10
+	// ErrorEntityVolume Entity
+	ErrorEntityVolume ErrorEntityType = 11
+)
 
-// NewErrorAndTimeNow returns instance of ErrorAndTime with the specified values
-func NewErrorAndTimeNow(errStr string) ErrorAndTime {
-	return ErrorAndTime{Error: errStr, ErrorTime: time.Now()}
+// ErrorEntity contains the device entity details
+type ErrorEntity struct {
+	EntityType ErrorEntityType // entity type
+	EntityID   string          // entity uuid
 }
 
-// ErrorAndTimeWithSource has an additional field "ErrorSourceType"
+// ErrorType has an additional field "ErrorSourceType"
 // which is used to selectively clear errors by calling IsErrorSource before
 // calling ClearErrorWithSource. See zedmanager and volumemgr for example use.
-type ErrorAndTimeWithSource struct {
-	ErrorSourceType string
-	Error           string
-	ErrorTime       time.Time
+type ErrorType struct {
+	ErrorSourceType     string
+	Error               string
+	ErrorTime           time.Time
+	ErrorSeverity       ErrorSeverity
+	ErrorRetryCondition string
+	ErrorEntities       []*ErrorEntity
```

Comment:
Isn't it easier if if we replace the Error string with a Error ErrorDescription where 
type ErrorDescription struct {
     ErrorString string
     ErrorSeverity ErrorSeverity
  	ErrorRetryCondition string
	ErrorEntities       []*ErrorEntity
}

Then we can 1) keep the errorandtime and errorandtimewithsource and 2) it is easier to pass the ErrorDescription from e.g., the downloaderStatus to VolumeStatus and 3) we can add more fields to the ErrorDescription over time.

Reply:
Sounds reasonable, done.